components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/IdentityKeyStoreResolver.java

```diff
@@ -0,0 +1,374 @@
+     * @return Primary key of Primary or tenant keystore
+     * @throws IdentityKeyStoreResolverException the exception in the IdentityKeyStoreResolver class
+     */
+    public Key getPrivateKey(String tenantDomain) throws IdentityKeyStoreResolverException {
+
+        int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+        if (privateKeys.containsKey(String.valueOf(tenantId))) {
+            return privateKeys.get(String.valueOf(tenantId));
+        }
+
+        KeyStoreManager keyStoreManager = KeyStoreManager.getInstance(tenantId);
+        Key privateKey;
+
+        try {
+            if (MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
+                privateKey = keyStoreManager.getDefaultPrivateKey();
+            } else {
+                String tenantKeyStoreName = buildTenantKeyStoreName(tenantDomain);
+                privateKey = keyStoreManager.getPrivateKey(tenantKeyStoreName, tenantDomain);
+            }
+        } catch (Exception e) {
+            throw new IdentityKeyStoreResolverException(
+                    "Error occurred when retrieving private key tenant: " + tenantDomain, e);
+        }
+
+        privateKeys.put(String.valueOf(tenantId), privateKey);
+        return privateKey;
+    }
+
+    /**
+     * Return Private Key of the Primary, tenant or custom keystore.
+     *
+     * @param tenantDomain      Tenant domain
+     * @param inboundProtocol   Inbound authentication protocol of the application
+     * @return Private Key of the Primary, tenant or custom keystore
+     * @throws IdentityKeyStoreResolverException the exception in the IdentityKeyStoreResolver class
+     */
+    public Key getPrivateKey(String tenantDomain, InboundProtocol inboundProtocol)
+            throws IdentityKeyStoreResolverException {
+
+        if (keyStoreMappings.containsKey(inboundProtocol)) {
+            if (MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain) ||
+                    keyStoreMappings.get(inboundProtocol).getUseInAllTenants()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Custom keystore configuration availble for " + inboundProtocol + " protocol.");
+                }
+
+                if (privateKeys.containsKey(inboundProtocol.toString())) {
+                    return privateKeys.get(inboundProtocol.toString());
+                }
+
+                String keyStoreName = buildCustomKeyStoreName(keyStoreMappings.get(inboundProtocol).getKeyStoreName());
+
+                try {
+                    int tenantId = IdentityTenantUtil.getTenantId(tenantDomain);
+                    KeyStoreManager keyStoreManager = KeyStoreManager.getInstance(tenantId);
+                    Key privateKey = keyStoreManager.getPrivateKey(keyStoreName, null);
+                    privateKeys.put(inboundProtocol.toString(), privateKey);
+                    return privateKey;
+                } catch (Exception e) {
```

Comment:
Catch specific exceptions not generic ones "Exception"

Reply:
Same as https://github.com/wso2/carbon-identity-framework/pull/5794#discussion_r1687747359